keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1159 @@
+
+#### kpt
+
+**Pattern**: list of GKNN (from ResourceGroup)
+
+Kpt uses a ResourceGroup CRD, and can register that CRD automatically.    The ResourceGroup contains a full list of GKNNs for all managed objects.  Kpt calls this full list of objects - including the names and namespaces - an “inventory”.  Each object gets an annotation `config.k8s.io/owning-inventory`, where that annotation corresponds to a label on the ResourceGroup `cli-utils.sigs.k8s.io/inventory-id`
+
+#### Google ConfigSync
+
+**Pattern**: list of GVKNN (from ResourceGroup)
+
+Distinct sets of synchronized resources are represented by RootSync / RepoSyncs, along with a ResourceGroup that has the full inventory.  Each object has some annotations that define membership, including the same `config.k8s.io/owning-inventory` as is used by kpt.    As with other solutions, following the “chain” from RootSync/RepoSync to managed objects is somewhat obscure.
+
+
+## Proposal
+
+A v2-prunable "apply set" is associated with an object on the cluster.   We define a set of standardized labels and annotations that identify the “parent object” of the apply set and the “member objects” of that parent.  We operate at the plumbing layer; we aim to support:
+
+- listing the parent objects efficiently (porcelain may expose this as listing groups of objects as managed by the various tools)
+- listing the member objects for a specific parent object efficiently (porcelain may use this for advanced diffing and pruning, and for presenting objects grouped by their higher-level aggregation)
+- basic apply-with-prune operations, where it creates or reuses a Secret in the cluster as the parent object.
+
+### Apply Set
+
+"Apply set" refers to a group of resources that are applied to the cluster by a tool. An apply set has a “parent” object of the tool’s preference. This “parent” object can be implemented using a ConfigMap, Secret, or a CRD of the tool’s choice.
+
+“ApplySet” is used to refer to the parent object in this design document, though the actual concrete resource on the cluster will typically be of a different Kind.  We might think of ApplySet as a “duck-type” based on the `applyset.k8s.io/id` label proposed here.
+
+### Member Object Labels
+
+Objects that are part of an ApplySet should carry a standardized label, with a key of:
+
+```yaml
+applyset.k8s.io/part-of: <applysetkey>
+```
+
+The `<applysetkey>` can be chosen essentially arbitrarily (subject to the limits of label values).
+
+>  <<[UNRESOLVED @justinsb @KnVerey]>>
+>
+> Should we identify a few patterns to avoid collisions?  Either uid:12345 or <toolname>:...
```

Comment:
I like the idea of allowing it to be arbitrary but suggesting `<toolname:...>`, e.g. `kubectl:<apply-set-name>`. I don't think we should recommend encoding uids, which aren't meaningful to humans. Plus imagine the scenario where the tracking object is deleted then re-created with the same name. I think the human's expectation would be that any previous children would continue to point to it.